internal/extproc/translator/gemini_helper.go

```diff
@@ -20,30 +28,485 @@ const (
+		}
+
+		// Generate a random ID for the tool call.
+		toolCallID := uuid.New().String()
+
+		toolCall := openai.ChatCompletionMessageToolCallParam{
+			ID:   toolCallID,
+			Type: "function",
+			Function: openai.ChatCompletionMessageToolCallFunctionParam{
+				Name:      part.FunctionCall.Name,
+				Arguments: string(args),
+			},
+		}
+
+		toolCalls = append(toolCalls, toolCall)
+	}
+
+	if len(toolCalls) == 0 {
+		return nil, nil
+	}
+
+	return toolCalls, nil
+}
+
+// toOpenAIUsage converts Gemini usage metadata to OpenAI usage.
+func toOpenAIUsage(metadata *genai.GenerateContentResponseUsageMetadata) openai.ChatCompletionResponseUsage {
+	if metadata == nil {
+		return openai.ChatCompletionResponseUsage{}
+	}
+
+	return openai.ChatCompletionResponseUsage{
+		CompletionTokens: int(metadata.CandidatesTokenCount),
+		PromptTokens:     int(metadata.PromptTokenCount),
+		TotalTokens:      int(metadata.TotalTokenCount),
+	}
+}
+
+// toLogprobs converts Gemini logprobs to OpenAI logprobs.
+func toLogprobs(logprobsResult genai.LogprobsResult) openai.ChatCompletionChoicesLogprobs {
+	if len(logprobsResult.ChosenCandidates) == 0 {
+		return openai.ChatCompletionChoicesLogprobs{}
+	}
+
+	content := make([]openai.ChatCompletionTokenLogprob, 0, len(logprobsResult.ChosenCandidates))
+
+	for i := 0; i < len(logprobsResult.ChosenCandidates); i++ {
+		chosen := logprobsResult.ChosenCandidates[i]
+
+		var topLogprobs []openai.ChatCompletionTokenLogprobTopLogprob
+
+		// Process top candidates if available.
+		if i < len(logprobsResult.TopCandidates) && logprobsResult.TopCandidates[i] != nil {
+			topCandidates := logprobsResult.TopCandidates[i].Candidates
+			if len(topCandidates) > 0 {
+				topLogprobs = make([]openai.ChatCompletionTokenLogprobTopLogprob, 0, len(topCandidates))
+				for _, tc := range topCandidates {
+					topLogprobs = append(topLogprobs, openai.ChatCompletionTokenLogprobTopLogprob{
+						Token:   tc.Token,
+						Logprob: float64(tc.LogProbability),
+					})
+				}
+			}
+		}
+
+		// Create token logprob.
+		tokenLogprob := openai.ChatCompletionTokenLogprob{
+			Token:       chosen.Token,
+			Logprob:     float64(chosen.LogProbability),
+			TopLogprobs: topLogprobs,
+		}
+
+		content = append(content, tokenLogprob)
+	}
+
+	// Return the logprobs.
+	return openai.ChatCompletionChoicesLogprobs{
+		Content: content,
+	}
+}
+
 func buildGCPModelPathSuffix(publisher, model, gcpMethod string) string {
 	pathSuffix := fmt.Sprintf("publishers/%s/models/%s:%s", publisher, model, gcpMethod)
 	return pathSuffix
 }
 
 // buildGCPRequestMutations creates header and body mutations for GCP requests
 // It sets the ":path" header, the "content-length" header and the request body.
-func buildGCPRequestMutations(path string, reqBody []byte) (*ext_procv3.HeaderMutation, *ext_procv3.BodyMutation) {
+func buildGCPRequestMutations(path *string, reqBody []byte) (*ext_procv3.HeaderMutation, *ext_procv3.BodyMutation) {
```

Comment:
people usually do not use a pointer to the string. This unnecessarily results in escaping the string header (a pair of the length of and pointer to the buffer) to heap. You can just use `len(path) != 0` to check if the string empty or not since I believe the empty path is invalid anyways.